tests/api_test.py

```diff
@@ -1765,6 +1765,12 @@ def f(a, b):
     with self.assertRaisesRegex(ValueError, msg):
       g(jnp.ones((1, 1)), b=1)
 
+  def test_primitive_jaxtype_error(self):
+    with self.assertRaisesRegex(
+        TypeError, "Argument .* of type .* is not a valid JAX type"):
+      lax.add(1, 'hi')
```

Comment:
I am surprised that the error in xla.abstractify is triggered here. I can't quite tell how. Probably by way of `core._valid_jaxtype` which is only called by `core._check_arg` but then I do not see how `lax.add` will call _check_arg. Is `lax.add` a compiled primitive? 

I see now, all arithmetic operators are compiled. This solution then does not work for the non-compiled lax primitives.

Reply:
Right yeah. In op-by-op mode the impls should check the types of their arguments, and any impl based on `xla.apply_primitive` needs to go through xla and hence `xla.abstractify`. In jaxpr tracing, the type checking should happen in the abstract eval rules.